### PR TITLE
New workers

### PR DESCRIPTION
**Pull request by magik6k**

### CODE REVIEW

File: cmd/lotus-bench/main.go [hunk 1]
```diff
@@ -137,11 +137,11 @@ func main() {
 				if err != nil {
 					return err
 				}
-				defer func() {
+				/*defer func() {
```
Comment (whyrusleeping) [inline cmd/lotus-bench/main.go]: remove or remove comment

File: cmd/lotus-seal-worker/main.go [hunk 2]
```diff
@@ -33,7 +49,7 @@ func main() {
 		Version: build.UserVersion,
 		Flags: []cli.Flag{
 			&cli.StringFlag{
-				Name:    "repo",
+				Name:    "workerrepo",
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: use the const?

File: cmd/lotus-seal-worker/main.go [hunk 3]
```diff
@@ -66,100 +76,208 @@ func main() {
...
+		&cli.BoolFlag{
+			Name:  "no-local-storage",
+			Usage: "don't use storageminer repo for sector storage",
+		},
+	},
 	Action: func(cctx *cli.Context) error {
-		/*	if !cctx.Bool("enable-gpu-proving") {
-				os.Setenv("BELLMAN_NO_GPU", "true")
-			}
+		if !cctx.Bool("enable-gpu-proving") {
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: where is this flag defined?

Comment (whyrusleeping) [reply cmd/lotus-seal-worker/main.go]: oh nvm, its on the root

File: cmd/lotus-seal-worker/main.go [hunk 4]
```diff
@@ -66,100 +76,208 @@ func main() {
...
+			if err := srv.Shutdown(context.TODO()); err != nil {
+				log.Errorf("shutting down RPC server failed: %s", err)
 			}
+			log.Warn("Graceful shutdown successful")
+		}()
 
-			nQueues := workers + transfers
-			var wg sync.WaitGroup
-			wg.Add(nQueues)
+		nl, err := net.Listen("tcp4", cctx.String("address"))
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: specifically tcp4?

File: cmd/lotus-storage-miner/storage.go [hunk 5]
```diff
@@ -103,3 +112,192 @@ var storageAttachCmd = &cli.Command{
...
+
+		local, err := nodeApi.StorageLocal(ctx)
+		if err != nil {
+			return err
+		}
+
+		for id, sectors := range st {
+			var u, s, c int
+			for _, decl := range sectors {
+				if decl.SectorFileType&sectorbuilder.FTUnsealed > 0 {
```
Comment (whyrusleeping) [inline cmd/lotus-storage-miner/storage.go]: this feels hacky, can't we just have a method that turns this into a sector file type, and we can switch on it?

File: markets/retrievaladapter/provider.go [hunk 6]
```diff
@@ -44,7 +44,17 @@ func (rpn *retrievalProviderNode) UnsealSector(ctx context.Context, sectorID uin
 	if err != nil {
 		return nil, err
 	}
-	return rpn.sealer.ReadPieceFromSealedSector(ctx, abi.SectorNumber(sectorID), sectorbuilder.UnpaddedByteIndex(offset), abi.UnpaddedPieceSize(length), si.Ticket.Value, *si.CommD)
+
+	mid, err := address.IDFromAddress(rpn.miner.Address())
+	if err != nil {
+		panic(err)
```
Comment (whyrusleeping) [inline markets/retrievaladapter/provider.go]: panic?

File: node/options.go [hunk 7]
```diff
@@ -138,8 +138,17 @@ func as(in interface{}, as interface{}) interface{} {
 
 	return reflect.MakeFunc(ctype, func(args []reflect.Value) (results []reflect.Value) {
 		outs := reflect.ValueOf(in).Call(args)
+
 		out := reflect.New(outType.Elem())
-		out.Elem().Set(outs[0])
+		if outs[0].Type().AssignableTo(outType.Elem()) {
```
Comment (whyrusleeping) [inline node/options.go]: reason for this change?

Comment (magik6k) [reply node/options.go]: This allow providing fx.In structs as implementations of interfaces - in this case this is for `api.Common` needed by `advmgr.New`

File: api/apistruct/struct.go [hunk 8]
```diff
@@ -189,6 +201,24 @@ type StorageMinerStruct struct {
...
+type WorkerStruct struct {
+	Internal struct {
+		// TODO: lower perms
+
+		Version func(context.Context) (build.Version, error) `perm:"admin"`
+
+		TaskTypes func(context.Context) (map[sealtasks.TaskType]struct{}, error) `perm:"admin"`
+		Paths     func(context.Context) ([]stores.StoragePath, error)            `perm:"admin"`
+		Info      func(context.Context) (api.WorkerInfo, error)                  `perm:"admin"`
+
```
Comment (lanzafame) [inline api/apistruct/struct.go]: @magik6k Is it possible for workers to be capable of performing the initial `pledge-sector` operation? I have noticed that it messes with a miner's ability to compute epost as there is no way to limit the number of gorountines used by `pledge-sector` and it starves epost

File: cmd/lotus-seal-worker/main.go [hunk 9]
```diff
@@ -66,100 +77,208 @@ func main() {
...
-				}()
+		go func() {
+			if err := nodeApi.WorkerConnect(ctx, "ws://"+cctx.String("address")+"/rpc/v0"); err != nil {
+				log.Errorf("Registering worker failed: %+v", err)
+				cancel()
+				return
 			}
+		}()
+
+		// todo go register
```
Comment (whyrusleeping) [inline cmd/lotus-seal-worker/main.go]: go register?

File: api/apistruct/struct.go (hunk 8 above)
Comment (magik6k) [reply api/apistruct/struct.go]: It's a fairly easy optimization, but I want to keep the initial thing minimal (If you can call this PR that, heh)

File: storage/sectorstorage/resources.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+	MultiThread bool
+	CanGPU      bool
+
+	BaseMinMemory uint64 // What Must be in RAM for decent perf (shared between threads)
+}
+
+const MaxCachingOverhead = 32 << 30
+
+var ResourceTable = map[sealtasks.TaskType]map[abi.RegisteredProof]Resources{
```
Comment (lanzafame) [inline storage/sectorstorage/resources.go]: Can we make this a configuration file so that tweaking doesn't have to be a recompilation of the binary?

File: storage/sectorstorage/manager.go [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+	// TODO: allow workers to fetch the sectors
+
+	best, err := m.index.StorageFindSector(ctx, sector, sectorbuilder.FTCache|sectorbuilder.FTSealed, true)
+	if err != nil {
+		return storage.SectorCids{}, xerrors.Errorf("finding path for sector sealing: %w", err)
+	}
+
+	candidateWorkers, _ := m.getWorkersByPaths(sealtasks.TTPreCommit2, best)
+	if len(candidateWorkers) == 0 {
+		return storage.SectorCids{}, xerrors.New("no suitable workers found")
```
Comment (whyrusleeping) [inline storage/sectorstorage/manager.go]: "no suitable workers found" seems like a good sentinel error to have

File: storage/sectorstorage/manager.go [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+	newWorkers chan *workerHandle
+	schedule   chan *workerRequest
+	workerFree chan WorkerID
+	closing    chan struct{}
+
+	schedQueue *list.List // List[*workerRequest]
+}
+
+func New(ls stores.LocalStorage, si stores.SectorIndex, cfg *sectorbuilder.Config, urls URLs, ca api.Common) (*Manager, error) {
+	ctx := context.TODO()
```
[the remark this answers is not in the document]
Comment (magik6k) [reply storage/sectorstorage/manager.go]: Because it's not wired to the root context with metrics. Will fix

File: storage/sectorstorage/resources.go (hunk 10 above)
Comment (magik6k) [reply storage/sectorstorage/resources.go]: There's nothing configurable about this table, it's just the resource use of sealing stages

Comment (magik6k) [reply storage/sectorstorage/resources.go]: (which is the same no matter what machine you have)

File: cmd/lotus-storage-miner/run.go [hunk 13]
```diff
@@ -60,6 +58,10 @@ var runCmd = &cli.Command{
 			return err
 		}
 
+		if _, _, err := ulimit.ManageFdLimit(); err != nil {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/lotus-storage-miner/run.go]: Added a flag allowing you to disable this (like in go-ipfs)

I've added it here because the default is usually too low to be useful, I'm yet to see a good reason to keep this limit low

File: storage/sectorstorage/manager.go [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+	m.remoteHnd.ServeHTTP(w, r)
+}
+
+func (m *Manager) SectorSize() abi.SectorSize {
+	sz, _ := m.scfg.SealProofType.SectorSize()
+	return sz
+}
+
+func (m *Manager) ReadPieceFromSealedSector(context.Context, abi.SectorID, sectorbuilder.UnpaddedByteIndex, abi.UnpaddedPieceSize, abi.SealRandomness, cid.Cid) (io.ReadCloser, error) {
+	panic("implement me")
```
Comment (whyrusleeping) [inline storage/sectorstorage/manager.go]: maybe we should just return an error for now?

File: storage/sectorstorage/manager.go [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+	if err != nil {
+		return xerrors.Errorf("finding sealed sector: %w", err)
+	}
+
+	candidateWorkers, _ := m.getWorkersByPaths(sealtasks.TTFinalize, best)
+	if len(candidateWorkers) == 0 {
+		return ErrNoWorkers
+	}
+
+	// TODO: Remove sector from sealing stores
```
Comment (whyrusleeping) [inline storage/sectorstorage/manager.go]: are these both still TODO?

File: storage/sectorstorage/roprov.go [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"golang.org/x/xerrors"
+)
+
+type readonlyProvider struct {
+	stor *stores.Local
+}
+
+func (l *readonlyProvider) AcquireSector(ctx context.Context, id abi.SectorID, existing sectorbuilder.SectorFileType, allocate sectorbuilder.SectorFileType, sealing bool) (sectorbuilder.SectorPaths, func(), error) {
+	if allocate != 0 {
```
Comment (whyrusleeping) [inline storage/sectorstorage/roprov.go]: magic number logic here is hard to follow

File: storage/sectorstorage/sectorutil/utils.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+	"github.com/filecoin-project/specs-actors/actors/abi"
+)
+
+func ParseSectorID(baseName string) (abi.SectorID, error) {
+	var n abi.SectorNumber
+	var mid abi.ActorID
+	read, err := fmt.Sscanf(baseName, "s-t0%d-%d", &mid, &n)
+	if err != nil {
+		return abi.SectorID{}, xerrors.Errorf(": %w", err)
```
Comment (whyrusleeping) [inline storage/sectorstorage/sectorutil/utils.go]: no error message?

File: storage/sectorstorage/stores/http_handler.go [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+	id, err := sectorutil.ParseSectorID(vars["id"])
+	if err != nil {
+		log.Error("%+v", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	ft, err := ftFromString(vars["type"])
+	if err != nil {
+		log.Error("%+v", err)
```
Comment (whyrusleeping) [inline storage/sectorstorage/stores/http_handler.go]: http 500 on these two too?

File: storage/sectorstorage/stores/index.go [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+	}
+	return nil
+}
+
+func (i *Index) StorageDeclareSector(ctx context.Context, storageId ID, s abi.SectorID, ft sectorbuilder.SectorFileType) error {
+	i.lk.Lock()
+	defer i.lk.Unlock()
+
+	for _, fileType := range pathTypes {
+		if fileType&ft == 0 {
```
Comment (whyrusleeping) [inline storage/sectorstorage/stores/index.go]: more magik bithacking >.>

File: storage/sectorstorage/stores/remote.go [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+	r.fetchLk.Lock()
+	defer r.fetchLk.Unlock()
+
+	paths, stores, done, err := r.local.AcquireSector(ctx, s, existing, allocate, sealing)
+	if err != nil {
+		return sectorbuilder.SectorPaths{}, sectorbuilder.SectorPaths{}, nil, xerrors.Errorf("local acquire error: %w", err)
+	}
+
+	for _, fileType := range pathTypes {
+		if fileType&existing == 0 {
```
Comment (whyrusleeping) [inline storage/sectorstorage/stores/remote.go]: `fileType.TellMeYourSecrets()`

File: storage/sectorstorage/stores/remote.go [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+		break
+	case 404:
+		return FsStat{}, errPathNotFound
+	case 500:
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return FsStat{}, xerrors.Errorf("fsstat: got http 500, then failed to read the error: %w", err)
+		}
+
+		return FsStat{}, xerrors.New(string(b))
```
Comment (whyrusleeping) [inline storage/sectorstorage/stores/remote.go]: default maybe?

**Comment by whyrusleeping** [comment]
also needs a rebase, and the tests are failing.

File: api/apistruct/struct.go [hunk 22]
```diff
@@ -175,12 +181,18 @@ type StorageMinerStruct struct {
 		SectorsRefs   func(context.Context) (map[string][]api.SealedRef, error)       `perm:"read"`
 		SectorsUpdate func(context.Context, abi.SectorNumber, api.SectorState) error  `perm:"write"`
 
-		/*		WorkerStats func(context.Context) (sectorbuilder.WorkerStats, error) `perm:"read"`
+		WorkerConnect func(context.Context, string) error                       `perm:"admin"` // TODO: worker perm
```
Comment (Kubuxu) [inline api/apistruct/struct.go]: //TODO

File: storage/sectorstorage/manager.go (hunk 14 above)
Comment (magik6k) [reply storage/sectorstorage/manager.go]: This is probably ok for now

File: storage/sectorstorage/manager.go (hunk 15 above)
Comment (magik6k) [reply storage/sectorstorage/manager.go]: Yes, I'm finishing this now

File: cmd/lotus-seal-worker/rpc.go [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+	"context"
+
+	"github.com/filecoin-project/specs-storage/storage"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/storage/sectorstorage"
+)
+
+type worker struct { // TODO: use advmgr.LocalWorker here
+	*sectorstorage.LocalWorker
```
Comment (Kubuxu) [inline cmd/lotus-seal-worker/rpc.go]: another TODO

File: storage/sectorstorage/stores/remote.go (hunk 20 above)
Comment (magik6k) [reply storage/sectorstorage/stores/remote.go]: https://github.com/filecoin-project/go-sectorbuilder/blob/feat/redo-interfaces/interface.go#L17

File: storage/sectorstorage/manager.go (hunk 15 above)
Comment (magik6k) [reply storage/sectorstorage/manager.go]: Actually will do as a separate PR

File: cmd/lotus-storage-miner/workers.go [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+
+		stats, err := nodeApi.WorkerStats(ctx)
+		if err != nil {
+			return err
+		}
+
+		st := make([]struct {
+			id uint64
+			api.WorkerStats
+		}, 0, len(stats))
```
Comment (Kubuxu) [inline cmd/lotus-storage-miner/workers.go]: You can just define this type within the function. This way you won't have to repeat type definition and it won't leak.

File: lib/lotuslog/levels.go [hunk 25]
```diff
@@ -14,5 +14,7 @@ func SetupLogLevels() {
 		logging.SetLogLevel("bitswap", "WARN")
 		//logging.SetLogLevel("pubsub", "WARN")
 		logging.SetLogLevel("connmgr", "WARN")
+		logging.SetLogLevel("advmgr", "DEBUG")
+		logging.SetLogLevel("stores", "DEBUG")
```
Comment (Kubuxu) [inline lib/lotuslog/levels.go]: Do we want it at debug or was it only for debugging?

Comment (magik6k) [reply lib/lotuslog/levels.go]: Let's keep it for now

File: storage/sectorstorage/resources.go [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+
+var FsOverheadFinalized = map[sectorbuilder.SectorFileType]int{
+	sectorbuilder.FTUnsealed: 10,
+	sectorbuilder.FTSealed:   10,
+	sectorbuilder.FTCache:    2,
+}
+
+type Resources struct {
+	MinMemory uint64 // What Must be in RAM for decent perf
+	MaxMemory uint64 // Mamory required (swap + ram)
```
Comment (Kubuxu) [inline storage/sectorstorage/resources.go]:
```suggestion
	MaxMemory uint64 // Memory required (swap + ram)
```

File: storage/sectorstorage/roprov.go (hunk 16 above)
Comment (Kubuxu) [reply storage/sectorstorage/roprov.go]: Yeah, constants would be great.

File: storage/sectorstorage/stores/http_handler.go [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+		return
```
Comment (Kubuxu) [inline storage/sectorstorage/stores/http_handler.go]: http error here too

File: storage/sectorstorage/stores/remote.go (hunk 21 above)
Comment (Kubuxu) [reply storage/sectorstorage/stores/remote.go]: yeah, something like `xerrors.Errorf("<description>: %s", b)`.

File: storage/sectorstorage/worker_remote.go [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	headers := http.Header{}
+	headers.Add("Authorization", "Bearer "+string(token))
+
+	wapi, close, err := client.NewWorkerRPC(url, headers)
+	if err != nil {
+		return nil, xerrors.Errorf("creating jsonrpc client: %w", err)
+	}
+	_ = close // TODO
```
Comment (Kubuxu) [inline storage/sectorstorage/worker_remote.go]: todo

File: api/apistruct/struct.go (hunk 22 above)
Comment (magik6k) [reply api/apistruct/struct.go]: Will leave for when we actually rework api permissions
